#!/usr/bin/perl

use Data::Dumper;
use FindBin;
use lib "$FindBin::Bin/../lib";
use Genbank qw(parse_genbank write_features_as_table);
use Subfunctions qw(parse_fasta merge_to_featuretable);

# a regionfile is the output of parse_blast.pl comparing the fastafile to the reference fasta file from genbank.pl
my $regionfile = shift;

# fastafile is the new fasta file to be aligned: same as fasta file used in parse_blast.pl
my $fastafile = shift;

# featurefile is the feature table file generated by genbank.pl
my $gbfile = shift;
my $outfile = shift;

my $gene_array = parse_genbank($gbfile);
open FEATURES_FH, ">", "$gbfile.features";
print FEATURES_FH write_features_as_table ($gene_array);
close FEATURES_FH;

my (undef, $fastaarray) = parse_fasta($fastafile);

# there should be only one key, so just one name.
my $name = @$fastaarray[0];

open FH, ">", "$outfile.tbl";
print FH merge_to_featuretable ($regionfile, "$gbfile.features", $name);
close FH;
# >Features Populus trichocarpa chloroplast, complete genome.
# 1	157033	source
# 			mol_type	genomic DNA
# 			organism	Populus trichocarpa
# 77	4	gene
# 			gene	trnH
# 77	4	tRNA
# 			gene	trnH
# 			product	tRNA-His

